PR For issue 22 - writing RK2 in cpp

Amey-kulkarni19 opened the pull request:
I have added C++ version of RK2 method in this branch along with a build file. @jcorbino @jbrzensk Please review it and let me know if any changes are required.

valeriabarra commented:
Hi @Amey-kulkarni19 , thank you for your contribution! 

Please note that the binary file is not needed. You do not need to check that in with git to the repository. 

If instead, you needed to make any local changes to, say, the [Makefile](https://github.com/csrc-sdsu/mole/blob/master/examples/cpp/Makefile) to be able to compile your example, than you want to check-in that file and push it upstream. 

Also, please make sure that your C++ example exactly reproduces the MATLAB one. You can do this for instance by running both examples and dumping the solution numerical values to files, then compare the files with a `diff` and take a screenshot of the comparison here in the PR conversation.

Thank you.

jbrzensk commented:
@Amey-kulkarni19  Appreciate the contribution!
We are very grateful for anything that makes the library more diverse.

You can include your Makefile, maybe stripping out some of your specific ARMA links, etc with:
git add -f Makefile, so we can see the build commands.

For comparison, because there is not a MATLAB example for RK2, maybe compare to the true solution (y=e^(1/2 t + sin(2t)/4))? You should really check my math there too! And then maybe, we could do an RK4 with the same kind of comparison?!?

Either print out the estimated and true values, the norm of the diff between your solution and the true solution (if it is nice and small), or include gnuplot showing the solution, both are good ideas.
system("gnuplot -persist 'plot' your data.txt")

aboada commented on examples/cpp/RK2.cpp:
```diff
@@ -0,0 +1,50 @@
+#include "mole.h"  // Include the MOLE library
```
Comments like this do not add any value and should be removed from the source code.

aboada commented on examples/cpp/RK2.cpp:
```diff
@@ -0,0 +1,50 @@
+double f(double t, double y);
+
+int main() {
+    double h = 0.1;               // Step size
```
If `h` is a constant, it should be declared as constexpr. I also recommend adhering to some style guidelines for naming conventions across files. For instance, you could use: https://google.github.io/styleguide/cppguide.html

aboada commented on examples/cpp/RK2.cpp:
You should have a more descriptive name for this file. `RK2.cpp` doesn't tell much. Also, please include comments that explain what kind of problem is being solved, conditions, methods used, etc.

valeriabarra replied:
Hi @aboada , thank you for your comments and review. The name of the file that @Amey-kulkarni19 chose for this example reflects the name of the [same RK2 example](https://github.com/csrc-sdsu/mole/blob/master/examples/matlab/RK2.m) in the `matlab` directory, so I think in this case it is fine to leave it like that. 

It is true that it is always best to leave documentation of what an example does, what the equations solved are, etc. So it would be good if @Amey-kulkarni19 could write a short description in the comments. 
Thank you!

valeriabarra replied:
It is ok to add a very simple description comment like there is in the corresponding Matlab example. 

```
// Solves ODE using explicit RK2 method
```

valeriabarra replied:
@Amey-kulkarni19 please do not resolve conversations. It is the role of the Reviewer to deem if comments have been addressed fully and resolve them. Here, for example, I still do not see the suggested change. Thank you

aboada replied:
> Hi @aboada , thank you for your comments and review. The name of the file that @Amey-kulkarni19 chose for this example reflects the name of the [same RK2 example](https://github.com/csrc-sdsu/mole/blob/master/examples/matlab/RK2.m) in the `matlab` directory, so I think in this case it is fine to leave it like that.
> 
> It is true that it is always best to leave documentation of what an example does, what the equations solved are, etc. So it would be good if @Amey-kulkarni19 could write a short description in the comments. Thank you!

In that case, they should be renamed to something more appropriate. What happens if you have another example that uses RK2?

aboada commented on examples/cpp/RK2.cpp:
```diff
@@ -0,0 +1,50 @@
+
+// Function definition for f(t, y)
+double f(double t, double y) {
+    return pow(sin(t), 2) * y;
```
should use `std::pow` and `std::sin` instead.

aboada commented on examples/cpp/RK2:
Please do not check in binaries in the repo.

Amey-kulkarni19 commented:
@aboada  @valeriabarra 

I have added GNU plot and pushed the makefile. I have also removed the build file.

Please review this and let me know.

Once this is merged I will create a new issue for RK4 and upload my code there.

valeriabarra commented:
Will close #22

aboada commented on examples/cpp/RK2.cpp:
```diff
@@ -0,0 +1,67 @@
+
+int main() {
+    constexpr double h = 0.1;      // Time step for the RK2 method
+    double t_start = 0.0;          // Start of the time interval
```
`t_start` and `t_end` can also be declared as `constexpr`

aboada commented on examples/cpp/RK2.cpp:
```diff
@@ -0,0 +1,67 @@
+
+    // Solve the ODE using the 2nd-order Runge-Kutta (RK2) method
+    for (int i = 0; i < n_steps - 1; ++i) {
+        double k1 = f(t(i), y(i));                            // Approximate slope at the beginning
```
Consider making k1 and k2 constants.

aboada commented on examples/cpp/RK2.cpp:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    // Open a pipe to send data and commands to GNUplot for visualization
+    FILE* gnuplot = popen("gnuplot -persist", "w");
```
I suggest you use `fstream/ofstream` instead.

valeriabarra commented on examples/cpp/Makefile:
```diff
@@ -46,5 +46,10 @@ elliptic1D: elliptic1D.cpp
 	$(CXX) $(CXXFLAGS) $(INCPATH) -o elliptic2D elliptic2D.cpp $(LIBS)
 
+RK2: RK2.cpp
+        $(CXX) $(CXXFLAGS) $(INCPATH) -o RK2 RK2.cpp $(LIBS)
```
Hi @Amey-kulkarni19 , thank you so much for your contribution!

You want to be very careful here. It looks like you have used spaces, while we want to use tab here. Otherwise, the `make` command will error with the following error message:

```
Makefile:50: *** missing separator (did you mean TAB instead of 8 spaces?).  Stop.
```

So you want:
```suggestion
	$(CXX) $(CXXFLAGS) $(INCPATH) -o RK2 RK2.cpp $(LIBS)
```

valeriabarra replied:
Also, you need to modify line 35 above where the `all` rule is defined and you want to add your executable to the list of files:

```
all: transport1D schrodinger1D elliptic1D elliptic2D RK2
```

valeriabarra commented on examples/cpp/RK2.cpp:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    // Write data to a file
+    std::ofstream data_file("data.txt");
+    if (!data_file) {
+        std::cerr << "Error: Failed to open data file.\n";
+        return 1;
+    }
```
I would not dump values to file, but to screen (standard output). 

However, you want to make sure that your C++ example exactly reproduces (or up to a good chunk of significant digits) the MATLAB example. So dumping values to file is a good debugging/verification practice. You can then do the same thing with the corresponding MATLAB file and compare the numerical values for correctness. 

However, in the version pushed to the PR, I would prefer to print out to standard output. Thank you!

valeriabarra commented:
> @Amey-kulkarni19 Appreciate the contribution! We are very grateful for anything that makes the library more diverse.
> 
> You can include your Makefile, maybe stripping out some of your specific ARMA links, etc with: git add -f Makefile, so we can see the build commands.
> 
> For comparison, because there is not a MATLAB example for RK2, maybe compare to the true solution (y=e^(1/2 t + sin(2t)/4))? You should really check my math there too! And then maybe, we could do an RK4 with the same kind of comparison?!?
> 
> Either print out the estimated and true values, the norm of the diff between your solution and the true solution (if it is nice and small), or include gnuplot showing the solution, both are good ideas. system("gnuplot -persist 'plot' your data.txt")

Hi @jbrzensk , thank you so much for your review/comments! Perhaps you misread. There actually is a corresponding [RK2 matlab example](https://github.com/csrc-sdsu/mole/blob/master/examples/matlab/RK2.m) on `master` already.

Amey-kulkarni19 commented:
@valeriabarra I have made following changes as per your suggestion. 
1. printing values instead of downloading a file
2.  added RK2 in all :
3. Replaced spaces with tab in makefile for RK2

Please let me know if anything else is needed.

Thank you for your guidance.

valeriabarra commented:
Please show that you have run a comparison with the corresponding  MATLAB and if you can show a screenshot of the comparison (see this message https://github.com/csrc-sdsu/mole/pull/27#discussion_r1884590367)

Amey-kulkarni19 commented:
@valeriabarra certainly!

Here's a screenshot for reference. The graph on left hand side is from .cpp file and the right hand side graph is .mat file
![comparison](https://github.com/user-attachments/assets/94f9e5df-7122-4b74-b00c-5b5c9e2b6852)

valeriabarra commented:
Hi @Amey-kulkarni19, thank you so much for posting those graphs. They look very good. However, it is always safer to look at the numerical values. Can you please check the numerical values of, say, only the last time step, and show the comparison of those? Thank you

Amey-kulkarni19 commented:
Output values in RK2.mat
![image](https://github.com/user-attachments/assets/d3e1a2c2-9ca9-4595-bbfb-c3a88c53f4ae)

Output values in RK2.cpp
![image](https://github.com/user-attachments/assets/ecde7fd5-37e9-4ae0-902a-83a1688a86f5)


As you can see, both values match perfectly.

valeriabarra commented on examples/cpp/Makefile:
```diff
@@ -1,4 +1,4 @@
-MOLE = ../../src/cpp
+MOLE = ../../src/mole_C++
```
Hi @Amey-kulkarni19 , the line above should not have changed. The actual path on `master` is `src/cpp`, so you should have kept 

```suggestion
MOLE = ../../src/cpp
```

Amey-kulkarni19 replied:
Oh!

I don't know how it changed. But I have corrected it.

valeriabarra commented:
Hi @Amey-kulkarni19 , thank you for your work. It looks good now. 

In the meantime, there were changes in the `master` branch that now are affecting one of the files you changed here in this PR (the `examples/cpp/Makefile`). Please update your branch to be on top of the master branch changes and incorporate/merge both the changes in `master` with your own changes so that you don't loose your changes and push again the updated version to this PR. If you are not sure how to do that please let me know.

Amey-kulkarni19 commented:
@valeriabarra I might need some guidance on this.

valeriabarra commented:
> @valeriabarra I might need some guidance on this.

Please see this comment that I left in another PR and the following comment https://github.com/csrc-sdsu/mole/pull/32#discussion_r1876858374

aboada commented:
@Amey-kulkarni19, thanks for the corrections!
